src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,375 @@
+         *
+         * @param parentPath The parent path to inject into the cache, can be null. Null will be treated as the root path.
+         * @param entry The entry to inject into the cache, must not be null.
+         */
+        private void injectIntoCache(@Nullable final Path parentPath, final ResourceCacheEntry entry) {
+            String pathEntry;
+            if (parentPath == null) {
+                //We have null, generally this only happens on the root directory.
+                pathEntry = "";
+            } else {
+                //Use the parent paths, we can not depend on a toString call here, since we need the / as separator.
+                pathEntry = Joiner.on("/").join(parentPath);
+            }
+
+            //Inject into the cache.
+            this.entriesByPathPrefix.putIfAbsent(pathEntry, new CopyOnWriteArrayList<>());
+            this.entriesByPathPrefix.get(pathEntry).add(entry);
+
+            //Recursively walk to the top, while preventing duplicate entries.
+            if (parentPath != null && !pathEntry.isEmpty()) {
+                this.injectIntoCache(parentPath.getParent(), entry);
+            }
+        }
+
+        /**
+         * Looks up resources in the given input path prefix as well as those that match the filter.
+         *
+         * @param inputPath The input path prefix to look in.
+         * @param filter The filter which the resources must match.
+         * @return A collection of resource location which match the given path prefix and who match the filter.
+         */
+        public Collection<ResourceLocation> getResources(final Path inputPath, final Predicate<ResourceLocation> filter) {
+            //Use the input path prefix, we can not depend on a toString call here, since we need the / as separator.
+            final String pathEntry = Joiner.on('/').join(inputPath);
+            if (!entriesByPathPrefix.containsKey(pathEntry)) {
+                //No cache entry found.
+                return Collections.emptyList();
+            }
+
+            //Since we inject into the cache recursively we can now just grab the map entry that is the prefix and loop over its values.
+            return entriesByPathPrefix.get(pathEntry).stream()
+                    .map(ResourceCacheEntry::resourceLocation)
+                    .filter(filter)
+                    .collect(Collectors.toList());
+        }
+
+        /**
+         * Indicates if the cache of this manager has been loaded.
+         *
+         * @return True if the cache has been loaded, false otherwise.
+         */
+        public boolean cacheLoaded() {
+            return cacheLoaded.get();
+        }
+
+        /**
+         * Stream element resource that combines a normalized path and a joined path using the "/" as separator.
+         */
+        private record PathWithLocationPath(Path path, String locationPath) { }
```

Comment:
This record could just be a local record rather than a private one as it is only used in a single method as a temporary helper to hold the data while mapping